### PR TITLE
Checksums

### PR DESCRIPTION
**Pull request by ilevyor**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,14 @@ Release
 *.iml
 tags
 
+### vscode ###
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
```
Comment (justinboswell) [inline .gitignore:18]: Remove these exceptions. NOTHING from .vscode should ever be committed.

File: include/aws/s3/aws_streaming_checksum.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+enum checksum_type {
+    Md5,
+    Sha1,
+    Sha256,
+    Crc32,
+    Crc32c,
+};
+
+AWS_S3_API
+struct aws_checksum *aws_sha256_new(struct aws_allocator *allocator);
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:17]: Just for completeness and good IDE integration results, go ahead and add:
```c
struct aws_checksum;
```

File: include/aws/s3/aws_streaming_checksum.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * Cleans up and deallocates checksum.
+ */
+AWS_S3_API
+void aws_checksum_destroy(struct aws_checksum *checksum);
+/**
+ * Updates the running checksum with to_checksum. this can be called multiple times.
+ */
+AWS_S3_API
+int aws_checksum_update(struct aws_checksum *checksum, const struct aws_byte_cursor *to_checksum);
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:47]: `to_checksum` should just be `data` or `buffer` or `or similar

File: include/aws/s3/aws_streaming_checksum.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_S3_CHECKSUM_H
+#define AWS_S3_CHECKSUM_H
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/common.h>
+#include <aws/s3/s3.h>
+
+enum checksum_type {
+    Md5,
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:9]: These should be caps, and prefix, e.g. `AWS_CHECKSUM_MD5`

File: include/aws/s3/aws_streaming_checksum.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+AWS_S3_API
+int aws_checksum_finalize(struct aws_checksum *checksum, struct aws_byte_buf *output, size_t truncate_to);
+
+/**
+ * Computes the md5 checksum over input and writes the digest output to 'output'.
+ * Use this if you don't need to stream the data you're checksuming and you can load
+ * the entire input to checksum into memory.
+ */
+AWS_S3_API
+int aws_md5_compute(
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:66]: This should be `aws_checksum_md5_compute`

File: include/aws/s3/aws_streaming_checksum.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * Computes the sha256 checksum over input and writes the digest output to 'output'.
+ * Use this if you don't need to stream the data you're checksuming and you can load
+ * the entire input to checksum into memory. If you specify truncate_to to something
+ * other than 0, the output will be truncated to that  number of bytes. For
+ * example if you want a SHA256 digest as the first 16 bytes, set truncate_to
+ * to 16. If you want the full digest size, just set this to 0.
+ */
+AWS_S3_API
+int aws_sha256_compute(
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:81]: `aws_checksum_sha256_compute`

File: include/aws/s3/aws_streaming_checksum.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * Computes the sha1 checksum over input and writes the digest output to 'output'.
+ * Use this if you don't need to stream the data you're checksuming and you can load
+ * the entire input to checksum into memory. If you specify truncate_to to something
+ * other than 0, the output will be truncated to that  number of bytes. For
+ * example if you want a SHA1 digest as the first 16 bytes, set truncate_to
+ * to 16. If you want the full digest size, just set this to 0.
+ */
+AWS_S3_API
+int aws_sha1_compute(
```
Comment (justinboswell) [inline include/aws/s3/aws_streaming_checksum.h:96]: `aws_checksum_sha1_compute`

**Review by justinboswell — CHANGES_REQUESTED** [review]

File: include/aws/s3/s3_streaming_checksum.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    AWS_CHECKSUM_MD5,
+    AWS_CHECKSUM_SHA1,
+    AWS_CHECKSUM_SHA256,
+    AWS_CHECKSUM_CRC32,
+    AWS_CHECKSUM_CRC32C,
+};
+*/
+#define AWS_SHA256_LEN 32
+/* #define AWS_SHA1_LEN 20 */
+#define AWS_CRC_LEN 4
```
Comment (justinboswell) [inline include/aws/s3/s3_streaming_checksum.h:23]: If these aren't needed/used, go ahead and delete them

File: include/aws/s3/s3_streaming_checksum.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+struct aws_checksum;
+
+typedef struct aws_checksum *(aws_checksum_new_fn)(struct aws_allocator *allocator);
```
Comment (justinboswell) [inline include/aws/s3/s3_streaming_checksum.h:27]: This is an implementation detail, doesn't need to be in the public header.

File: include/aws/s3/s3_streaming_checksum.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+ * to 16. If you want the full digest size, just set this to 0.
+ */
+/*
+AWS_S3_API
+int aws_checksum_sha1_compute(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *input,
+    struct aws_byte_buf *output,
+    size_t truncate_to);
+*/
```
Comment (justinboswell) [inline include/aws/s3/s3_streaming_checksum.h:100]: Don't check in commented out code, add this back when you add SHA-1 support.

**Review by justinboswell — COMMENTED** [review]
Go ahead and knock out the SHA-1 stuff, and this is good to go

File: source/crc32.c [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    checksum->good = false;
+    /* checksum->impl has type (void *) to match the rest of the API, but we are storing as a uintptr, and using it as
+     * an int to avoid mem allocation */
+    uintptr_t crc_value = (uintptr_t)checksum->impl;
+    const uint32_t crc = (uint32_t)crc_value;
+    /* changed to BE in response to @graebm hesitant comment */
+    /* is this conditional necessary? */
+    return aws_byte_buf_write_be32(output, crc) ? AWS_OP_SUCCESS : AWS_OP_ERR;
```
Comment (graebm) [inline source/crc32.c:108]: tweak comment, we know that big-endian is needed
`/* Write out bytes in big endian order */`

File: include/aws/s3/s3_streaming_checksum.h [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+#define AWS_SHA256_LEN 32
+/* #define AWS_SHA1_LEN 20 */
+#define AWS_CRC_LEN 4
+
+struct aws_checksum;
+
+typedef struct aws_checksum *(aws_checksum_new_fn)(struct aws_allocator *allocator);
+
+AWS_S3_API
+struct aws_checksum *aws_checksum_sha256_new(struct aws_allocator *allocator);
```
[the remark this answers is not in the document]
Comment (ilevyor) [reply include/aws/s3/s3_streaming_checksum.h:30]: @justinboswell

**Pull request closed without merging**